runners/direct-java/src/main/java/org/apache/beam/runners/direct/ExecutorServiceParallelExecutor.java

```diff
@@ -277,18 +277,35 @@ private void shutdownIfNecessary(State newState) {
       return;
     }
     LOG.debug("Pipeline has terminated. Shutting down.");
-    pipelineState.compareAndSet(State.RUNNING, newState);
+
     // Stop accepting new work before shutting down the executor. This ensures that thread don't try
     // to add work to the shutdown executor.
-    serialExecutorServices.invalidateAll();
-    serialExecutorServices.cleanUp();
-    parallelExecutorService.shutdown();
-    executorService.shutdown();
+    try {
+      serialExecutorServices.invalidateAll();
+    } catch (final RuntimeException re) {
+      LOG.error(re.getMessage(), re);
```

Comment:
Why are we swallowing failures here? Especially `RuntimeExceptions`

Reply:
Cause they are done in threads without reject handler so it leads to the same except now we call all cleanups and dont leak if anything happens.

Reply:
Have you encountered failures as a result of us throwing out of these methods? If so, what was it? We should control all of the things we're trying to shutdown, and they shouldn't throw any exceptions unless our code is buggy. All of this code centers around rejecting future work, and none of the objects should throw while updating their internal states to reflect that.

I'm not opposed to putting the whole method content (at least, the parts that have actual code) within a `try/finally`.

Reply:
All lines can fail (Error thrown but we never catch them or just the internal impl which can throw an exception). Since all lines are important I prefered to protect them this way to avoid to leak threads or miss invalidations. try/finally was the alternative but it requires one per line which makes the code super nested for a poor gain IMHO so I went with this option.

Reply:
`serialExecutorService.shutdown()` can't throw an exception unless `queue.clear()` throws an exception, and I am not convinced that `ConcurrentLinkedQueue` will ever throw an exception
`parallelExecutorService.shutdown()` can never throw an exception

I'm not hugely concerned about throwing an exception out of `executorService.shutdown()`, or either of the `cache` cleanup chain.

I would find the single `try/finally` block to be significantly more readable than chained `try/catch` blocks.

Reply:
Hmm no, when changing the state it could throw an exception. Hughly unlikely but possible and preventing next libe is an issue for this whole block. Not much possible, I agree, but a corner case it doesnt cost much to manage which can save us from leaks and weird behaviors.

Reply:
I'm not sure what you mean by "changing the state". More importantly, exceptions can't "just happen" in the way errors could; they represent a specific type of failure, which none of these methods are permitted to exhibit.

Reply:
the executor manages a state, when calling shutdown it mutates this state and has some conditions where it can fail. It is very unlikely today - I didnt fully check it - but since the direct runner needs more and more custom pools it will be more likely soon - when the change will be done. This is why I prefer to be extract cautious here because if a day it breaks it will be very hard to find and to fix for just a few try/catch now.

Reply:
Swallowing the exceptions and logging them isn't really a good way to propagate the failure modes of shutting down executors, as the Pipeline will appear to shut down in a healthy state, which isn't really appropriate, given that we've potentially knowingly leaked some resources.

I don't think we'll need many more (or any more) custom pools.

Reply:
Also, I still don't know why we can't (for propagation reasons) move these into the same `try` block as `registry.cleanup()`. When pipeline cleanup fails in _this_ context (the `DirectRunner`), I find it entirely acceptable to propagate that back as a system failure which has caused the pipeline to fail.

Reply:
Putting them all in a single block is very fragile since all single line can fail with a runtime exception - even if unlikely today. I will aggregate the errors in a list and throw an illegalstateexception as aggregator with all messages for now. I will also ensure we hit a leaf state before throwing this exception to keep a valid state in the pipeline result.